core/meterer/onchain_state.go

```diff
@@ -12,82 +12,111 @@ import (
-	GetActiveReservations(ctx context.Context, blockNumber uint32) (map[string]core.ActiveReservation, error)
-	GetActiveReservationsByAccount(ctx context.Context, blockNumber uint32, accountID string) (core.ActiveReservation, error)
-	GetOnDemandPayments(ctx context.Context, blockNumber uint32) (map[string]core.OnDemandPayment, error)
-	GetOnDemandPaymentByAccount(ctx context.Context, blockNumber uint32, accountID string) (core.OnDemandPayment, error)
+	RefreshOnchainPaymentState(ctx context.Context, tx *eth.Transactor) error
+	GetActiveReservations(ctx context.Context) (map[string]core.ActiveReservation, error)
+	GetActiveReservationByAccount(ctx context.Context, accountID string) (core.ActiveReservation, error)
+	GetOnDemandPayments(ctx context.Context) (map[string]core.OnDemandPayment, error)
+	GetOnDemandPaymentByAccount(ctx context.Context, accountID string) (core.OnDemandPayment, error)
+	GetOnDemandQuorumNumbers(ctx context.Context) ([]uint8, error)
 }
 
 type OnchainPaymentState struct {
 	tx *eth.Transactor
 
-	ActiveReservations map[string]core.ActiveReservation
-	OnDemandPayments   map[string]core.OnDemandPayment
+	ActiveReservations    map[string]core.ActiveReservation
+	OnDemandPayments      map[string]core.OnDemandPayment
+	OnDemandQuorumNumbers []uint8
 }
 
 func NewOnchainPaymentState(ctx context.Context, tx *eth.Transactor) (OnchainPaymentState, error) {
-	activeReservations, onDemandPayments, err := CurrentOnchainPaymentState(ctx, tx)
+	blockNumber, err := tx.GetCurrentBlockNumber(ctx)
+	if err != nil {
+		return OnchainPaymentState{}, err
+	}
+
+	quorumNumbers, err := tx.GetRequiredQuorumNumbers(ctx, blockNumber)
 	if err != nil {
-		return OnchainPaymentState{tx: tx}, err
+		return OnchainPaymentState{}, err
 	}
 
 	return OnchainPaymentState{
-		tx:                 tx,
-		ActiveReservations: activeReservations,
-		OnDemandPayments:   onDemandPayments,
+		tx:                    tx,
+		ActiveReservations:    make(map[string]core.ActiveReservation),
+		OnDemandPayments:      make(map[string]core.OnDemandPayment),
+		OnDemandQuorumNumbers: quorumNumbers,
 	}, nil
 }
 
-// CurrentOnchainPaymentState returns the current onchain payment state (TODO: can optimize based on contract interface)
-func CurrentOnchainPaymentState(ctx context.Context, tx *eth.Transactor) (map[string]core.ActiveReservation, map[string]core.OnDemandPayment, error) {
+// RefreshOnchainPaymentState returns the current onchain payment state (TODO: can optimize based on contract interface)
+func (pcs *OnchainPaymentState) RefreshOnchainPaymentState(ctx context.Context, tx *eth.Transactor) error {
 	blockNumber, err := tx.GetCurrentBlockNumber(ctx)
 	if err != nil {
-		return nil, nil, err
+		return err
 	}
 
-	activeReservations, err := tx.GetActiveReservations(ctx, blockNumber)
-	if err != nil {
-		return nil, nil, err
+	accountIDs := make([]string, 0, len(pcs.ActiveReservations))
+	for accountID := range pcs.ActiveReservations {
```

Comment:
where does `pcs.ActiveReservations` first get populated?

Reply:
`ActiveReservations` gets populated on the request basis. When a request comes in, we would check if its account id exists in this map, and query the contract if not. We can't populate it at initialization because we don't want to store a list of account IDs on the contract and there's no function to take the keys from the contract storage mapping